File: java/src/Mcp.java
package hermes;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;

class McpPlugin {
    String name;
    Process process;
    OutputStream stdin;
    InputStream stdout;
}

public class Mcp {
    private final Map<String, McpPlugin> plugins = new ConcurrentHashMap<>();

    public void load(String name, String command, String[] args) throws Exception {
        List<String> commandList = new ArrayList<>();
        commandList.add(command);
        commandList.addAll(Arrays.asList(args));

        ProcessBuilder builder = new ProcessBuilder(commandList);
        Process process = builder.start();

        McpPlugin plugin = new McpPlugin();
        plugin.name = name;
        plugin.process = process;
        plugin.stdin = process.getOutputStream();
        plugin.stdout = process.getInputStream();

        plugins.put(name, plugin);
        System.out.println("[hermes:mcp] Loaded plugin: " + name);
    }

    public void unload(String name) {
        McpPlugin plugin = plugins.remove(name);
        if (plugin != null) {
            plugin.process.destroyForcibly();
            System.out.println("[hermes:mcp] Unloaded plugin: " + name);
        }
    }
}

File: rust/src/mcp.rs
use std::collections::HashMap;
use std::process::{Command, Child, Stdio};
use std::io::{self, Write, Read};
use std::sync::RwLock;

pub struct Plugin {
    pub name: String,
    process: Child,
}

pub struct PluginManager {
    plugins: RwLock<HashMap<String, Plugin>>,
}

impl PluginManager {
    pub fn new() -> Self {
        Self {
            plugins: RwLock::new(HashMap::new()),
        }
    }

    pub fn load(&self, name: &str, command: &str, args: &[&str]) -> io::Result<()> {
        let mut child = Command::new(command)
            .args(args)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .spawn()?;

        let plugin = Plugin {
            name: name.to_string(),
            process: child,
        };

        let mut plugins = self.plugins.write().unwrap();
        plugins.insert(name.to_string(), plugin);
        println!("[hermes:mcp] Loaded plugin: {}", name);

        Ok(())
    }

    pub fn unload(&self, name: &str) -> io::Result<()> {
        let mut plugins = self.plugins.write().unwrap();
        if let Some(mut plugin) = plugins.remove(name) {
            plugin.process.kill()?;
            println!("[hermes:mcp] Unloaded plugin: {}", name);
        }
        Ok(())
    }
}
